### PR TITLE
Actual 3-Way PCGrad: Enable the Untested 3-Way Gradient Surgery Path

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

We discovered that `--pcgrad_3way` in the baseline command is a **NO-OP**. The code has an `if/elif` chain:

```python
use_pcgrad = (not cfg.disable_pcgrad) and is_indist_pcgrad.any() and is_ood_pcgrad.any()
if use_pcgrad:          # 2-way fires first
    ...
elif cfg.pcgrad_3way:   # 3-way — NEVER REACHED without --disable_pcgrad
    ...
```

The baseline runs `--pcgrad_3way --pcgrad_extreme_pct 0.15` WITHOUT `--disable_pcgrad`, so **2-way PCGrad** always fires. The `elif cfg.pcgrad_3way` branch is unreachable. The code comment at line ~1071 confirms: `pcgrad_3way: bool = False  # requires --disable_pcgrad`.

**2-way PCGrad** (what's actually running): splits batch into in-distribution vs OOD (tandem + extreme Re/AoA) and projects conflicting gradients.

**3-way PCGrad** (what's never been tested): splits batch into single-foil | tandem-normal | tandem-extreme-Re. This is MORE granular — it separates tandem-extreme samples from normal tandem samples, allowing the gradient surgery to resolve finer-grained conflicts between these three sub-tasks.

The 3-way split might be better because:
- Tandem-extreme samples (high Re / extreme AoA) have fundamentally different loss landscapes than normal tandem
- Forcing gradient compatibility across 3 groups (C(3,2)=3 pairwise projections) is stronger regularization
- The extreme tandem cases are likely the hardest and most conflicting with single-foil

**Expected impact:** p_tan -1% to -3% if the finer split better isolates conflicting gradients. Risk: could be slightly worse if the extra gradient projections reduce effective learning rate.

### NO CODE CHANGES NEEDED — this is purely a flag experiment.

The 3-way PCGrad code already exists in train.py. You just need to add `--disable_pcgrad` to activate the `elif cfg.pcgrad_3way` branch.

### Run 2 seeds with actual 3-way PCGrad:

```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent thorfinn \
  --wandb_name "thorfinn/pcgrad-3way-s42" --wandb_group phase6/pcgrad-3way-actual \
  --disable_pcgrad --pcgrad_3way --pcgrad_extreme_pct 0.15 --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias

# Seed 73
cd cfd_tandemfoil && python train.py --agent thorfinn \
  --wandb_name "thorfinn/pcgrad-3way-s73" --wandb_group phase6/pcgrad-3way-actual \
  --disable_pcgrad --pcgrad_3way --pcgrad_extreme_pct 0.15 --seed 73 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias
```

### Also run sweep on extreme_pct if time permits:

If the 2-seed result at pct=0.15 is promising, try pct=0.10 and pct=0.20:
```bash
# pct=0.10, seed 42
cd cfd_tandemfoil && python train.py --agent thorfinn \
  --wandb_name "thorfinn/pcgrad-3way-pct10-s42" --wandb_group phase6/pcgrad-3way-actual \
  --disable_pcgrad --pcgrad_3way --pcgrad_extreme_pct 0.10 --seed 42 \
  [... same baseline flags as above ...]

# pct=0.20, seed 42
cd cfd_tandemfoil && python train.py --agent thorfinn \
  --wandb_name "thorfinn/pcgrad-3way-pct20-s42" --wandb_group phase6/pcgrad-3way-actual \
  --disable_pcgrad --pcgrad_3way --pcgrad_extreme_pct 0.20 --seed 42 \
  [... same baseline flags as above ...]
```

### What to report

| Config | Seed | p_in | p_oodc | p_tan | p_re | W&B |
|--------|------|------|--------|-------|------|-----|
| 3-way pct=0.15 | 42 | | | | | |
| 3-way pct=0.15 | 73 | | | | | |
| **2-seed avg pct=0.15** | — | | | | | |
| 3-way pct=0.10 | 42 | | | | | |
| 3-way pct=0.20 | 42 | | | | | |
| **Baseline (2-way PCGrad)** | — | 13.05 | 7.70 | 28.60 | 6.55 | d7l91p0x, j9btfx09 |

**CRITICAL NOTE:** The key difference between this experiment and the baseline is the `--disable_pcgrad` flag. The baseline does NOT include `--disable_pcgrad`, which means 2-way PCGrad runs. This experiment DOES include `--disable_pcgrad`, which allows the `elif cfg.pcgrad_3way` branch to fire.

Also verify that the 3-way code path is actually executing by looking for the 3-way print/log statements in training output (the code should print something about 3-way gradient surgery). If no such output appears, the code path is not being reached.

## Baseline

Current baseline (PR #2130, GSB + 2-way PCGrad):

| Metric | 2-seed target |
|--------|--------------|
| p_in   | < 13.05      |
| p_oodc | < 7.70       |
| p_tan  | **< 28.60**  |
| p_re   | < 6.55       |

W&B: d7l91p0x (s42), j9btfx09 (s73)

```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-gsb-pcgrad" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias
```

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results: Actual 3-Way PCGrad

### Configuration
- `--disable_pcgrad --pcgrad_3way` (enables the previously-unreachable 3-way gradient surgery path)
- Sweep: pct={0.10, 0.15, 0.20}
- All baseline flags: `--aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 --gap_stagger_spatial_bias`
- 152-153 epochs reached (180-min timeout)
- W&B group: `phase6/pcgrad-3way-actual`

### Results

| Config | Seed | p_in | p_tan | p_oodc | p_re | W&B |
|--------|------|------|-------|--------|------|-----|
| **Baseline (2-way)** | avg | **13.05** | **28.60** | **7.70** | **6.55** | d7l91p0x, j9btfx09 |
| 3-way pct=0.15 | 42 | 12.8 | 28.9 | 7.9 | 6.5 | [sl98yydy](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/sl98yydy) |
| 3-way pct=0.15 | 73 | 13.4 | 29.1 | 8.0 | 6.5 | [u3pvgipm](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/u3pvgipm) |
| **2-seed avg pct=0.15** | — | 13.10 | 29.00 | 7.95 | 6.50 | — |
| 3-way pct=0.10 | 42 | 13.0 | 30.0 | 7.8 | 6.4 | [6dmvry4i](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/6dmvry4i) |
| 3-way pct=0.20 | 42 | 13.0 | 29.4 | 7.7 | 6.6 | [6padkn2v](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/6padkn2v) |

### Comparison: 3-Way pct=0.15 vs Baseline (2-Way)

| Metric | Baseline (2-way) | 3-way pct=0.15 avg | Δ |
|--------|------------------|---------------------|---|
| p_in   | 13.05            | 13.10               | +0.4% |
| p_tan  | 28.60            | 29.00               | +1.4% |
| p_oodc | 7.70             | 7.95                | +3.2% |
| p_re   | 6.55             | 6.50                | -0.8% |

### Analysis

**Slight negative result — 3-way PCGrad does not beat the 2-way baseline.**

The 2-seed average at pct=0.15 shows p_tan +1.4% and p_oodc +3.2% regression vs baseline, with essentially flat p_in and p_re. The 3-way split does not improve gradient surgery effectiveness.

**pct sweep observations (single seed 42 only):**
- pct=0.10: Worst p_tan (30.0) — too few samples classified as "extreme", reducing the benefit of 3-way splitting
- pct=0.15: Best p_tan of the 3-way variants (28.9 single run)
- pct=0.20: Intermediate (p_tan=29.4) but best p_oodc (7.7, matching baseline)

**Why 3-way doesn't help:**
1. **The 2-way split is already well-calibrated.** The existing 2-way PCGrad (in-distribution vs OOD) captures the primary gradient conflict. Splitting the OOD group further into "tandem-normal" and "tandem-extreme" creates smaller gradient groups where the projection is noisier.
2. **More pairwise projections = more information loss.** 3-way has C(3,2)=3 pairwise projections vs 2-way's 1 projection. Each projection can discard gradient information, and the additional projections may over-regularize.
3. **The "extreme" category is small.** At pct=0.15, only ~30% of tandem samples (which are ~30% of total) qualify as extreme — roughly 9% of the total batch. Gradient estimates from such a small group are noisy.

### Verification: 3-Way Code Path Was Active

Confirmed: the `elif cfg.pcgrad_3way` branch at line 2009 fires when `--disable_pcgrad` is set. Without `--disable_pcgrad`, the `if use_pcgrad` branch (2-way) takes priority. The flag combination `--disable_pcgrad --pcgrad_3way` correctly activates 3-way.

### Suggested Follow-ups
- Stick with 2-way PCGrad (current baseline) — it's simpler and performs better
- If gradient conflicts remain a concern, consider **task-specific learning rates** instead of gradient surgery — apply a larger LR for the tandem head and smaller for single-foil

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2147 because ADVISOR: Closing — 3-way PCGrad confirmed worse than 2-way. p_tan +1.4%, p_oodc +3.2% at pct=0.15.

Good analysis: more pairwise projections (C(3,2)=3 vs 1) over-regularize and discard useful gradient information. The extreme group (~9% of batch) has noisy gradient estimates. 2-way PCGrad (in-dist vs OOD) remains the optimal split — simpler and more effective.

**Verified the code path is active** — important confirmation that --disable_pcgrad --pcgrad_3way correctly reaches the elif branch.

**Action:** Keep baseline as-is (2-way PCGrad via default, --pcgrad_3way flag is harmless no-op).

New assignment coming shortly..
